### PR TITLE
feat(middleware): support creation of standalone functions

### PR DESCRIPTION
**Pull request by TheEdoRan**
# Proposed changes

Code in this PR requires context to be an object, it extends it by default, and enables creation of standalone middleware functions via built-in `experimental_createMiddleware` utility.

## Related issue(s) or discussion(s)

re #222

### CODE REVIEW

**Review by mcalhoun — COMMENTED** [review]
Thanks for this work! I left a few comments on the deepMerge function (I've got a bunch of experience in this area and there are lots of foot guns!

File: packages/next-safe-action/src/utils.ts [hunk 1]
```diff
@@ -2,6 +2,22 @@ export const DEFAULT_SERVER_ERROR_MESSAGE = "Something went wrong while executin
 
 export const isError = (error: unknown): error is Error => error instanceof Error;
 
+export const deepMerge = (obj1: object, obj2: object) => {
+	for (const key of Object.keys(obj2)) {
+		const k = key as keyof typeof obj2;
+		// eslint-disable-next-line
+		if (typeof obj2[k] === "object" && Object.hasOwn(obj1, k)) {
+			// @ts-expect-error
```
[the remark this answers is not in the document]
Comment (TheEdoRan) [reply packages/next-safe-action/src/utils.ts:10]: Thank you for reviewing the code. Would it be better to use lodash merge function in your opinion?

[the remark this answers is not in the document]
Comment (mcalhoun) [reply packages/next-safe-action/src/utils.ts:10]: I definitely think using a 3rd-party library would be better, although I don't think lodash has been well-maintained recently? Maybe consider [ramda](https://github.com/ramda/ramda)?

[the remark this answers is not in the document]
Comment (TheEdoRan) [reply packages/next-safe-action/src/utils.ts:10]: The thing is: ramda is published as a whole library, while lodash provides [lodash.merge](https://www.npmjs.com/package/lodash.merge) as a standalone package, which is obviously much smaller in size. [deepmerge](https://www.npmjs.com/package/deepmerge) or [deepmerge-ts](https://www.npmjs.com/package/deepmerge-ts) also appear to be viable options. What do you think?

**Comment by mcalhoun** [comment]
If you use the tsup tree shaking option, it should package only the code
you end up using in the final build output. Lodasb certainly works, but the
code is about 5 years old. I don't have experience with the other
libraries.

**Comment by TheEdoRan** [comment]
@mcalhoun I chose to use [deepmerge-ts](https://www.npmjs.com/package/deepmerge-ts), which seems to be an up to date and more performant variant of deepmerge. Apparently, if you install the package as `devDependency`, tsup takes care of bundling it in the js output, which is nice.

**Pull request merged**